### PR TITLE
WAL-G storage tool copy command

### PR DESCRIPTION
**Pull request by reshke**
Storage tool that provides cross-storage objects copying. Also, additional argument `--prefix` allow to filter only objects with names having specific prefix in them. This comes in handy when only some folders need to be copied.

Example:
```
./main/gp/wal-g --config ./conf.yaml st copy -f ./conf.yaml -t ./conf2.yaml -p segments_005/seg1/basebackups_005/yezzey
```
Another example:
```
 ./main/gp/wal-g --config ./conf1.yaml st copy --from ./conf1.yaml --to ./conf2.yaml -d -e
```

### Database name
Wal-g provides support for many databases, please write down name of database you uses.

# Pull request description

### Describe what this PR fix
// problem is ...

### Please provide steps to reproduce (if it's a bug)
// it can really help

### Please add config and wal-g stdout/stderr logs for debug purpose

also you can use WALG_LOG_LEVEL=DEVEL for logs collecting
<details><summary>If you can, provide logs</summary>
<p>
```bash
any logs here
```
</p>
</details>

### CODE REVIEW

File: internal/splitmerge/splitmerge_test.go [hunk 1]
```diff
@@ -34,9 +34,11 @@ func (b *BufferCloser) Close() error {
 	return nil
 }
 
-//               ┌─> copy data per 1 byte    ─>┐
+//	┌─> copy data per 1 byte    ─>┐
+//
 // data ─> split ├─> copy data per ... bytes ─>├─> merge
-//               └─> copy data per 42 bytes  ─>┘
+//
```
Comment (reshke) [inline internal/splitmerge/splitmerge_test.go:40]: this is `make fmt` artefact

Comment (ostinru) [reply internal/splitmerge/splitmerge_test.go:40]: Lol. I found a workaround - add emoji:
```
// 🙈               ┌─> copy data per 1 byte    ─>┐
// 🙉 data ─> split ├─> copy data per ... bytes ─>├─> merge
// 🙊               └─> copy data per 42 bytes  ─>┘
```

PS: or cdots =)
```
// ·               ┌─> copy data per 1 byte    ─>┐
// · data ─> split ├─> copy data per ... bytes ─>├─> merge
// ·               └─> copy data per 42 bytes  ─>┘
```

**Comment by reshke** [comment]
Hi, i have updated this pr, this now contains logic to decrypt files from source storage and encrypt files before putting them into target folder, because source and target config may contain different encryption settings. 

Note, than compress/uncompress logic is not needed here, because data copy can be done without this!

**Comment by reshke** [comment]
There is problem with `--config` file actually. You can-not run `wal-g` without specifying it, while it is not needed for this exact command.

File: internal/storagetools/copy_object_handler.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+	"github.com/wal-g/tracelog"
+	"github.com/wal-g/wal-g/internal"
+	"github.com/wal-g/wal-g/internal/copy"
+	"github.com/wal-g/wal-g/pkg/storages/storage"
+
+	"github.com/wal-g/wal-g/internal/crypto"
+	"github.com/wal-g/wal-g/utility"
+)
+
+// CompressAndEncrypt compresses input to a pipe reader. Output must be used or
```
Comment (reshke) [inline internal/storagetools/copy_object_handler.go:16]: comment need to be updated. tbd

**Review by usernamedt — CHANGES_REQUESTED** [review]

File: cmd/common/st/copy_object.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+	fromConfigFile string
+	toConfigFile   string
+	prefix string
+
+	decryptSource bool
+	encryptTarget bool
+)
+
+// catObjectCmd represents the catObject command
+var copyObjectCmd = &cobra.Command{
```
[the remark this answers is not in the document]
Comment (reshke) [reply cmd/common/st/copy_object.go:49]: yeah, done

File: docker/st_tests/scripts/tests/storage_tool_tests.sh [hunk 4]
```diff
@@ -46,9 +46,48 @@ wal-g st rm testfolder/testfile.br
...
+AWS_ENDPOINT: "$AWS_ENDPOINT"
+WALE_S3_PREFIX: "s3://storagetoolsbucket_target"
+AWS_S3_FORCE_PATH_STYLE: $AWS_S3_FORCE_PATH_STYLE
+AWS_ACCESS_KEY_ID: "$AWS_ACCESS_KEY_ID"
+AWS_SECRET_ACCESS_KEY: "$AWS_SECRET_ACCESS_KEY"
+EOH
+
+wal-g --config ./conf1.yaml st copy --from ./conf1.yaml --to ./conf2.yaml
+
+# Should get empty storage after file copy
```
Comment (usernamedt) [inline docker/st_tests/scripts/tests/storage_tool_tests.sh:92]:
```suggestion
# Should not have an empty storage after the file copy
```

Comment (reshke) [reply docker/st_tests/scripts/tests/storage_tool_tests.sh]: Thanks!

**Review by usernamedt — APPROVED** [review]

**Review by teem0n — APPROVED** [review]

**Pull request merged**